In [ ]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from matplotlib.ticker import MultipleLocator
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

<h1>Блок 1</h1>

<h3>Вопрос 1</h3>

**Таблица 1**
<br>
$$ARPPU = 3.14$$ <br>
$$ARPU = 0.33$$ <br>
$$DPU = 23415$$ <br>

$$Payers = \frac{ARPU}{ARPPU} = \frac{0.33}{3.14} = 0.1051 = 10.51\%$$ <br>
<br>
$$DAU * payers = DPU \Rightarrow DAU = \frac {23415}{0.1051} = 222797$$ <br>
<br>
$$Revenue = DAU * ARPU = 222797*0.33 = 72523.1$$


**Таблица 2**
<br>
$$Revenue = 59400$$ <br>
$$DPU = 7425$$ <br>
$$Payers = 5.5\%$$ <br>
$$DAU * payers = DPU \Rightarrow DAU = \frac {7425}{0.055} = 135000$$  <br>
$$Revenue = DAU * ARPU \Rightarrow ARPU = \frac{59400}{135000} = 0.44 $$ <br>
$$ARPU = ARPPU * Payers \Rightarrow ARPPU = \frac{0.44}{0.055} = 8$$

Во второй игре процент платящих пользователей меньше, но больше доход с каждого. Сравним отношения. <br>
<br>
$\frac{Payers1}{Payers2} = \frac{0.1051}{0.055} = 1.91$
<br><br>
$\frac{ARPPU2}{ARPPU1} = \frac{8}{3.14} = 2.54$ <br>
<br>
Как видим, доход с пользователя во второй игре больше в ~2.5 раза, в то время как платят меньше только в ~1.9 раза. Поэтому вторая игра лучше. <br>
В целом, это же подтверждает и доход с пользователя по всем пользователям (ARPU), у второй игры он выше (прямое следствие из посчитанного выше).

<h3>Вопрос 2</h3>

- Возможно, стоит обратить внимание на сложность уровней (особенно если были обновления). Они могут быть слишком сложными, от чего игроки устают.
- Возможно, наоборот, игра проходится слишком легко, и пользователи успеют полностью пройти её. Также оцениваем сложность.
- Возможно, стало больше рекламы (или она стала слишком нерелевантной), что может надоедать пользователям.
- В любом случае, стоит смотреть на субъекты последних обновлений (системы бонусов, дизайн, уровни). Ключ может быть в них.

<h3>Вопрос 3</h3>

Высокий LT говорит о том, что пользователи заинтересованы в игре, и очень лояльны к ней. Я считаю, что временная распродажа игровой валюты только укрепит лояльность пользователей, и не только поможет быстро заработать больше, но и поможет ещё дольше удерживать пользователей в игре, ведь они получат доступ к новым расширенным возможностям, закупившись валютой. <br><br>
Кроме того, возможен кейс, когда игрок проводил много времени в игре, но не мог себе позволить покупку. Распродажа поможет ему с покупкой, станет стимулом к вложению денег, что опять же выгодная для всех ситуация.

<h3>Вопрос 4</h3>

Ситуация похожа на вопрос 1, с двумя таблицами. <br><br>
Пусть $x$ и $y$ - срдений чек и количество платящих. <br>
Тогда прибыль была $xy$, а стала $0.9072xy$. <br><br>
Фактически, сейчас мы в минусе, однако, так как обновление успешно с точки зрения играбельности, потенциально довольные пользователи могут обеспечить рост приложения, который позволит перекрыть текущие потери в 9.28% и заработать значительно больше. <br><br>
Кроме того, ряд давних пользователей может ещё только привыкать в новой фактуре приложения, поэтому они ещё не решаются на пополнение баланса игры. Однако, так как геймплей большинству по душе, они ещё могут захотеть что-то приобрести, а значит падение количества платежей временное.
<br><br>
Обновление можно считать успешным.

<h3>Вопрос 5</h3>

In [32]:
df1_5 = pd.read_csv('ex1.5.csv', sep = '\t')
df1_5['ARPU1'] = (df1_5.Hint1 + df1_5.Hint2*2.5) / df1_5['Player count']
df1_5['ARPU2'] = (df1_5.Hint1*0.5 + df1_5.Hint2*0.2) / df1_5['Player count']
df1_5['player_pct_change'] = round(100*df1_5['Player count'].pct_change().shift(-1), 2)
df1_5.head(3)

Level  Hint1  Hint2  Player count     ARPU1     ARPU2  player_pct_change
0     36    292     93         23589  0.022235  0.006978              -5.12
1     37    977    683         22381  0.119945  0.027930              -4.43
2     38    465    392         21390  0.067555  0.014535              -1.76

Посмотрим на ситуацию 1.

In [33]:
df1_5[['Level', 'ARPU1', 'player_pct_change']].sort_values(by = 'ARPU1', ascending = False).head(5)

Level     ARPU1  player_pct_change
6      42  0.290600             -28.10
13     49  0.285171              -9.82
9      45  0.230191              -1.64
10     46  0.192144             -14.42
11     47  0.188109              -3.42

Видим, что с отрывом высокие показатели доходности у уровней 42 и 49 (будем считать разницу в 0.005 незначительной), однако уровень 42 значительно сильнее «отпугивает» игроков, аж 28% не возвращается на следующий уровень. А значит, **уровень 49** можно считать лучшим.

In [34]:
df1_5[['Level', 'ARPU1', 'player_pct_change']].sort_values(by = 'ARPU1', ascending = True).head(5)

Level     ARPU1  player_pct_change
0      36  0.022235              -5.12
8      44  0.029222              -4.64
2      38  0.067555              -1.76
12     48  0.074388             -12.86
5      41  0.104581              -3.52

Аналогично, худшим уровнем можно считать **36**: приносит меньше всех денег, но ещё и отпугивает людей сильнее, чем, например, 44. <br>

Посмотрим на ситуацию 2.

In [35]:
df1_5[['Level', 'ARPU2', 'player_pct_change']].sort_values(by = 'ARPU2', ascending = False).head(5)

Level     ARPU2  player_pct_change
6      42  0.077080             -28.10
10     46  0.059390             -14.42
13     49  0.054283              -9.82
4      40  0.041845              -9.72
11     47  0.041704              -3.42

С отрывом самый доходный **уровень — 42**. Однако он слишком сильно отпугвает людей в дальнейшем. Возможно, в этом плане уровень 46 даже лучше. 

In [36]:
df1_5[['Level', 'ARPU2', 'player_pct_change']].sort_values(by = 'ARPU2', ascending = True).head(5)

Level     ARPU2  player_pct_change
0      36  0.006978              -5.12
8      44  0.006994              -4.64
2      38  0.014535              -1.76
12     48  0.022732             -12.86
7      43  0.024295              -2.50

Как и в ситуации 1, худшим уровнем можно назвать **36**.

<h1>Блок 2</h1>

In [4]:
df2 = pd.read_csv('ex2.csv', sep = '\t')
df2['quality'] = df2.item_id % 100
df2['type'] = df2.item_id // 10000
df2

item_id    From  Count        Date  Level  VIP  quality  type
0       10004  Store1      2   9/29/2017      2    1        4     1
1       20004  Store1     10   9/29/2017      5    1        4     2
2       10001  Store2     64   9/29/2017      1    0        1     1
3       20002  Store2     25   9/29/2017      1    1        2     2
4       30002  Store2     46   9/29/2017      2    0        2     3
...       ...     ...    ...         ...    ...  ...      ...   ...
4990    30009   Sale1    144  10/23/2017      6    1        9     3
4991    30005   Sale1    160  10/23/2017      7    1        5     3
4992    20009   Sale1    176  10/23/2017      8    1        9     2
4993    10007   Sale1    120  10/23/2017      9    1        7     1
4994    20005   Sale1     96  10/23/2017     10    1        5     2

[4995 rows x 8 columns]

Видим проседающие по продажам предметов 29 и 30 августа.

In [5]:
df2.groupby('Date')['Count'].agg('sum').sort_values(ascending = False)

Date
10/19/2017    19733
10/18/2017    17319
10/20/2017    13036
10/5/2017     12591
10/6/2017     12491
10/21/2017    11904
10/13/2017    11552
10/7/2017     11509
10/14/2017    10771
10/12/2017    10456
10/9/2017     10374
10/8/2017     10286
10/22/2017    10082
10/11/2017     9858
10/17/2017     9627
10/10/2017     9535
10/15/2017     9201
10/16/2017     9199
10/4/2017      6926
10/2/2017      6849
10/3/2017      5904
10/23/2017     5729
10/1/2017      5120
9/29/2017      2084
9/30/2017      1785
Name: Count, dtype: int64

19.10.2017 нетипично хорошо покупали предметы не на распродаже. Отчасти, конечно, это связано с тем, что в этот день впринципе было больше продаж. <br><br>
Строка с индексом 4026 — единственная для предметов, купленных не на распродаже, входящая в квантиль 0.995 по числу покупок.

In [6]:
df2[df2.From.isin(['Store1', 'Store2'])].sort_values(by = 'Count', ascending = False).head(3)

item_id    From  Count        Date  Level  VIP  quality  type
4026    30006  Store1    510  10/19/2017     11    1        6     3
4045    20006  Store1    480  10/19/2017     11    1        6     2
4064    20008  Store1    370  10/19/2017     11    1        8     2

In [7]:
df2[(df2.Count > df.Count.quantile(0.99)) & (df2.From.isin(['Store1', 'Store2']))]

item_id    From  Count        Date  Level  VIP  quality  type
4026    30006  Store1    510  10/19/2017     11    1        6     3

Обратим внимание на строки с максимальным количеством продаж.

In [8]:
df2[df2.Count == df2.Count.max()]

item_id   From  Count        Date  Level  VIP  quality  type
3910    30007  Sale1   2288  10/18/2017     11    1        7     3
3955    10009  Sale1   2288  10/18/2017     11    1        9     1

<h1>Блок 3</h1>

Разницы, когда тратить конкретные бонусы нет, мы и так, и так получим одинаковое количество опыта. <br>
- Сначала посмотрим, сколько опыта наберёт игрок просто заходя в приложение 50 дней: <br>
50\*100 + 7*5500 = 43500 <br>
- Этого хватит на покупку 145 уровней (остаток — 0 у.е.)
- Прокачавшись до 145 уровня пользователь получит:
  - 8 яблок (400 у.е.)
  - 7 апельсинов (525 у.е.)
  - 8 груш (1040 у.е.)
- Итого у нас на балансе 1965 у.е.
- Самый выгодный спосок покупки уровней, как ни странно, 5 уровней за 120 у.е. (24 у.е. за 1 уровень).
- Покупаем на все, 16*5 = 80 уровней. В остатке 45 у.е.
- Итого у нас 225 уровень. Значит получили ещё:
  - 3 яблока, 3 апельсина, 2 груши, суммарно 635 у.е.
- На балансе 680 у.е. Покупаем 25 уровней за 600. В остатке 80 у.е., уровень 250.
- Получили ещё 1 грушу, 1 яблоко, 1 апельсин, суммарно 255 у.е. В остатке 335 у.е.
- Очевидно, приближаемся к моменту, когда уже не сможем купить уровни. Два варианта:
  - Снова тратим 240 на 10 уровней, в остатке 95, уровень 260. Продаём полученную грушу, на балансе 225 у.е. Покупаем 5 уровней, 265 уровень, остаток 105. Стоп.
  - Покупаем 11 уровней за 275. Уровень 261. Остаток 60. Продаём полученную грушу, баланс становится 190. Покупаем 5 уровней, уровень становится 266. На балансе 70. Стоп.
- Как видим, вторая стратегия выгоднее. 
- Однако, это значит, что в прошлом мы могли два раза взять не 10 уровней за 240, а два раза 11 за 275 (цель оптимизации - конечный уровень). 
- Предположим, мы просчитали всё заранее, а значит сделали это на моменте, когда у нас был баланс 680, мы потратили на 70 больше, но получили на 2 уровня больше (это не влияет на получение новых фруктов - до 270 не дойдём всё равно). 

**Итого**
- Уровень: 268
- Яблок — 12, апельсинов — 11, груш — 12
- Потрачено = заработано = 46485
- На балансе: 0

<h1>Блок 4</h1>

In [9]:
df4 = pd.read_csv('ex4.csv', sep = '\t')
df4['percent_win'] = round(df4.win/df4.play, 3)
df4['percent_lose'] = round(df4.lose / df4.play, 3)
df4['percent_quit'] = round(df4.quit / df4.play, 3)
df4['not_win_nearly'] = round(df4.nearly / (df4.lose + df4.quit), 3)
df4['gold_per_user'] = round(df4.gold / df4.play, 3)
df4

id    play    win  ...  percent_quit  not_win_nearly  gold_per_user
0      1  100253  58197  ...         0.365           0.742          0.459
1      2  117368  72262  ...         0.035           0.240          0.228
2      3   77872  44299  ...         0.019           0.175          0.404
3      4   89036  44538  ...         0.015           0.520          0.055
4      5   59219  36378  ...         0.013           0.992          0.596
..   ...     ...    ...  ...           ...             ...            ...
97    98   27463  25621  ...         0.004           0.413          0.774
98    99   13788  13101  ...         0.003           0.761          0.874
99   100    5428   5328  ...         0.001           0.810          0.373
100  101    1132   1120  ...         0.001           0.000          0.456
101  102     113     90  ...         0.027           0.522          0.619

[102 rows x 12 columns]

**1. Самый сложный уровень.** <br>

Отсортируем по проценту побед (возрастание) и проценту близких к победе (возрастание). 
<br> Самые «плохие» показатели у уровня **4**. <br>
По сравнению с 11-м многие не то, что проиграли, но не были даже близки к победе.

In [10]:
df4[['id', 'percent_win', 'percent_lose', 'not_win_nearly']].sort_values(by = ['percent_win', 'not_win_nearly'], ascending = (True, False)).head(5)

id  percent_win  percent_lose  not_win_nearly
10  11        0.498         0.484           0.709
3    4        0.500         0.485           0.520
17  18        0.568         0.418           0.571
2    3        0.569         0.413           0.175
48  49        0.577         0.408           0.951

**2. Самый лёгкий уровень.** <br>

Отсортируем по проценту побед и проценту близких к победе. Самые высокие показатели у уровня **71**.

In [11]:
df4[['id', 'percent_win', 'not_win_nearly']].sort_values(by = ['percent_win', 'not_win_nearly'], ascending = False).head(5)

id  percent_win  not_win_nearly
70    71        0.990           0.394
61    62        0.989           0.095
100  101        0.989           0.000
52    53        0.988           0.146
71    72        0.985           0.394

**3. Самый лучший уровень.** <br>

Уровень **20** не только приносит больше всего денег с одной игры, но и имеет хорошие показатели проходимости и выходов, а значит не раздражает игроков и не провоцирует их покинуть игру. <br>
В целом, уровень **99** ему эквивалентен: игроки тратят незначительно меньше, но реже выходят и чаще побеждают. А значит, он приятнее.

In [12]:
df4[['id', 'play', 'gold_per_user', 'percent_quit', 'percent_win']].sort_values(by = 'gold_per_user', ascending = False).head(5)

id   play  gold_per_user  percent_quit  percent_win
19  20  23002          0.900         0.004        0.938
98  99  13788          0.874         0.003        0.950
12  13  11723          0.850         0.006        0.924
81  82   1204          0.844         0.002        0.948
35  36  26673          0.839         0.002        0.932

**4. Часто вынуждающие к выходу.**

Просто сортируем по проценту выходов. Это уровень **22**.

In [13]:
df4[['id', 'play', 'percent_quit']].sort_values(by = 'percent_quit', ascending = False).head(5)

id    play  percent_quit
21  22  269891         0.378
45  46  114043         0.376
54  55  187419         0.374
37  38   60971         0.373
29  30   80881         0.372

**5. Уровень, где тратят больше всего.** <br>

Сортируем по золоту, потраченному на человека. Получаем уровень **20**.

In [14]:
df4[['id', 'play', 'gold_per_user']].sort_values(by = 'gold_per_user', ascending = False).head(5)

id   play  gold_per_user
19  20  23002          0.900
98  99  13788          0.874
12  13  11723          0.850
81  82   1204          0.844
35  36  26673          0.839

<h1>Блок 5</h1>

**Вопрос 1.** <br>
- Резкие скачки. <br> 
- Можно определить допустивые уровни максимального отклонения от среднего вверх и вниз. <br> 
В частности, например, если распределение заболевших по дням близко к нормальному, то через +/- $3\sigma$. 


**Вопрос 2.** <br> 
- Статистический тест для проверки нулевой гипотезы о том, что коэффициент корреляции между событиями пунктов 1 и 2 равен нулю. Соответственно, если гипотеза уверенно отвергается на выбранном уровне значимости, то связь есть.
- Просто посчитать коэффициент корреляции через .corr(). Если он сильно отличен от нуля, то связь, очевидно, есть.

**Вопрос 3.** <br> 
- Анализ временного ряда и построение прогноза по уже имеющимся данным.

**Вопрос 4.** <br>
- Возьмём две группы, тех, кто принимал плацебо, и тех, кто принимал разработанную вакцину. Сравним статистическим тестом (t-критерий Стьюдента) нулевую гипотезу о равенстве сроков выздоровления. <br>
- Если нулевую гипотезу не получилось отвергнуть на удовлетворяющем нас уровне значимости, то мы не можем сделать вывод, что разработанная вакцина не имеет эффекта, так как срок выздоровления не изменился.
- Также может быть применим статистический подход с проверкой гипотезы о равенстве вероятностей, если, например, какое-то количество людей от z-вируса не выздоравливает вовсе.  